simulation/agent.py

```diff
@@ -271,9 +324,22 @@ def on_vision_update(self) -> None:
         pass
 
     @abstractmethod
-    def on_noise(self, noise: 'world.NoiseEvent') -> None:
-        """ Noise handler, will be called before `on_tick` """
-        pass
+    def on_noise(self, noises: List['world.NoiseEvent']) -> None:
+        """ Noise handler, that checks, if there a noise event occurs and where it is perceived  """
+        if self._is_deaf:
+            return
+        
+        # noises
+        for noise in noises:
+            distance = math.sqrt((noise.location.x-self.location.x)**2+(noise.location.y-self.location.y)**2)
+            if(distance < noise.radius and noise.source != self):
+                dot = noise.location.x*self.location.x + noise.location.y*self.location.y      # dot product
+                det = noise.location.x*self.location.y - noise.location.y*self.location.x      # determinant
+                true_angle = math.atan2(det, dot)  
+                uncertainty = 10
+                perceived_angle = np.random.normal(true_angle, uncertainty)
+                return perceived_angle
```

Comment:
You have an early exit here, you only process the first valid noise and just skip the rest instead of returning a list